scipy/special/_basic.py

```diff
@@ -3207,61 +3237,64 @@ def stirling2(N, K, *, exact=True):
-        snsk_vals = defaultdict(int)
-        for pair in [(0, 0), (1, 1), (2, 1), (2, 2)]:
-            snsk_vals[pair] = 1
-        n_old, n_row = 2, [0, 1, 1]
-        # for each pair in the min-heap, calculate the value, store for later
-        while nk_pairs:
-            n, k = heappop(nk_pairs)
-            if n < 2 or k > n or k <= 0:
-                continue
-            elif k == n or k == 1:
-                snsk_vals[(n, k)] = 1
-                continue
-            elif n != n_old:
-                num_iters = n - n_old
-                while num_iters > 0:
-                    n_row.append(1)
-                    # traverse from back to remove second row
-                    for j in range(len(n_row)-2, 1, -1):
-                        n_row[j] = n_row[j]*j + n_row[j-1]
-                    num_iters -= 1
-                snsk_vals[(n, k)] = n_row[k]
-            else:
-                snsk_vals[(n, k)] = n_row[k]
-            n_old, n_row = n, n_row
-        # for each pair in the map, fetch the value, and populate the array
-        it = np.nditer(
-            [N, K, None],
-            ['buffered', 'refs_ok'],
-            [['readonly'], ['readonly'], ['writeonly', 'allocate']],
-            op_dtypes=[object, object, object],
-        )
-        with it:
-            while not it.finished:
-                it[2] = snsk_vals[(int(it[0]), int(it[1]))]
-                it.iternext()
-            output = it.operands[2]
-            # If N and K were both scalars, convert output to scalar.
-            if output_is_scalar:
-                output = output.take(0)
-            return output
-    else:  # this branch will house future Temme approx
-        raise NotImplementedError()
+    # make a min-heap of unique (n,k) pairs
+    N, K = asarray(N), asarray(K)
+    if not np.issubdtype(N.dtype, np.integer):
+        raise TypeError("Argument `N` must contain only integers")
+    if not np.issubdtype(K.dtype, np.integer):
+        raise TypeError("Argument `K` must contain only integers")
+    nk_pairs = list(
+        set([(n.take(0), k.take(0))
+             for n, k in np.nditer([N, K], ['refs_ok'])])
+    )
+    heapify(nk_pairs)
+    # base mapping for small values
+    snsk_vals = defaultdict(int)
+    for pair in [(0, 0), (1, 1), (2, 1), (2, 2)]:
+        snsk_vals[pair] = 1
+    # for each pair in the min-heap, calculate the value, store for later
+    n_old, n_row = 2, [0, 1, 1]
+    while nk_pairs:
+        n, k = heappop(nk_pairs)
+        if n < 2 or k > n or k <= 0:
+            continue
+        elif k == n or k == 1:
+            snsk_vals[(n, k)] = 1
+            continue
+        elif not exact:
+            # NOTE: here we allow np.uint via casting to double types prior to
+            # passing to private ufunc dispatcher. All dispatched functions
+            # take double type for (n,k) arguments and return double.
+            snsk_vals[(n, k)] = _stirling2_inexact(np.double(n),np.double(k))
```

Comment:
I think you should just have an `if not exact` just after the input validation and before creating `nk_pairs` and then return either`_stirling2_inexact(np.double(N), np.double(K))` or `_stirling2_inexact(N.astype(float), K.astype(float))`. From a readability perspective I prefer the latter, but the performance will be the same and you can just use whichever you like better.

I guess you wanted to still benefit from computing everything in one go for the `exact=False` case? I haven't profiled things; did you find your current approach works better in practice when using array input? My thought is that eventually we're going to use a higher order expansion that gives accuracy closer to machine precision and we want user to have access to vectorized computation that performs the inner loops in C++.

Reply:
> I guess you wanted to still benefit from computing everything in one go for the `exact=False` case? I haven't profiled things; did you find your current approach works better in practice when using array input? 

I think that was what I was going for at the time but yeah that ignores all the vectorization we get with the ufunc machinery.  I didn't profile anything so I'll switch to the single `if not exact` branch and call the ufunc. 

> My thought is that eventually we're going to use a higher order expansion that gives accuracy closer to machine precision and we want user to have access to vectorized computation that performs the inner loops in C++.

I agree.

Reply:
commit `5b70fd7` has changes to address this comment-thanks for catching this!